### PR TITLE
JDBC SchemaTransform implementation. Need to break out into a separat…

### LINKED ISSUE #21408: Convert JDBC SchemaIO to SchemaTransform
Here's the SchemaIO version:

[sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcSchemaIOProvider.java](https://github.com/apache/beam/blob/2a2107af70dea8c34743cb009be7904b6497858b/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcSchemaIOProvider.java)

Imported from Jira [BEAM-14033](https://issues.apache.org/jira/browse/BEAM-14033). Original Jira may contain additional context.
Reported by: laraschmidt.
Subtask of issue #21406

### PR DESCRIPTION
**Pull request by byronellis**
SchemaTransform implementation for JdbcIO. Follows the pattern of various other SchemaTransforms, splitting Read and Write into separate transforms and using a URN pattern consistent with other transforms such as Kafka. Configurations are largely intended to match the existing SchemaIO implementation (which appears that it might have issues with usernames and passwords, but would handle that in a separate PR)

Fixes #21408

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Abacn** [comment]
R: @ahmedabu98 for schema transform PRs

**Review by ahmedabu98 — CHANGES_REQUESTED** [review]
Looks good! Just a few nits about how we check for "null" configuration fields here.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+    JdbcReadSchemaTransformConfiguration config;
+
+    public JdbcReadSchemaTransform(JdbcReadSchemaTransformConfiguration config) {
+      this.config = config;
+    }
+
+    protected JdbcIO.DataSourceConfiguration dataSourceConfiguration() {
+      JdbcIO.DataSourceConfiguration dsConfig =
+          JdbcIO.DataSourceConfiguration.create(config.getDriverClassName(), config.getJdbcUrl())
```
[the remark this answers is not in the document]
Comment (byronellis) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java:65]: Updated validation to check for blank strings

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+          String query = config.getReadQuery();
+          if (query == null) {
+            query = String.format("SELECT * FROM %s", config.getLocation());
+          }
+          JdbcIO.ReadRows readRows =
+              JdbcIO.readRows()
+                  .withDataSourceConfiguration(dataSourceConfiguration())
+                  .withQuery(query);
+          Short fetchSize = config.getFetchSize();
+          if (fetchSize != null) {
```
[the remark this answers is not in the document]
Comment (byronellis) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java]: Yup, I also modified username and password to account for this behavior as well.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  @Override
+  protected @UnknownKeyFor @NonNull @Initialized Class<JdbcReadSchemaTransformConfiguration>
+      configurationClass() {
+    return JdbcReadSchemaTransformConfiguration.class;
+  }
+
+  @Override
+  protected @UnknownKeyFor @NonNull @Initialized SchemaTransform from(
+      JdbcReadSchemaTransformConfiguration configuration) {
+    if (configuration.getReadQuery() != null && configuration.getLocation() != null) {
```
[the remark this answers is not in the document]
Comment (byronellis) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java]: Started getting a little overloaded so I moved stuff to a validate method and applied changes.

**Review by ahmedabu98 — CHANGES_REQUESTED** [review]
Thank you for the changes! Just a couple more nits

File: sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProviderTest.java [hunk 4]
```diff
@@ -65,28 +65,68 @@ public static void beforeClass() throws Exception {
...
+          JdbcReadSchemaTransformProvider.JdbcReadSchemaTransformConfiguration.builder()
+              .setDriverClassName("ClassName")
+              .setJdbcUrl("JdbcUrl")
+              .build()
+              .validate();
+        });
+  }
 
+  @Test
+  public void testInvalidWriteSchemaOptions() {
```
[the remark this answers is not in the document]
Comment (byronellis) [reply sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProviderTest.java]: Good catch! Moved.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java [hunk 5]
```diff
@@ -164,6 +158,26 @@ public abstract static class JdbcReadSchemaTransformConfiguration implements Ser
     @Nullable
     public abstract Boolean getOutputParallelization();
 
+    public void validate() throws IllegalArgumentException {
+      if ("".equals(getDriverClassName())) {
+        throw new IllegalArgumentException("JDBC Driver class name cannot be blank.");
+      }
+      if ("".equals(getJdbcUrl())) {
+        throw new IllegalArgumentException("JDBC URL cannot be blank");
+      }
```
[the remark this answers is not in the document]
Comment (byronellis) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java]: Sure thing. Done.

**Review by ahmedabu98 — APPROVED** [review]
LGTM, thanks for implementing this SchemaTransform!

@Abacn can you merge this when tests pass?

**Comment by Abacn** [comment]
Looks like jenkins was down shortly and tests not ran

**Comment by Abacn** [comment]
retest this please

**Comment by byronellis** [comment]
Took the opportunity to add a couple of documentation blurbs :-)

**Comment by byronellis** [comment]
@Abacn All tests are now passing

**Comment by Abacn** [comment]
Thanks @byronellis @ahmedabu98 !

**Pull request merged**